libpod/image/image.go

```diff
@@ -149,43 +210,344 @@ func (i *Image) ID() string {
+		} else {
+			names = append(names, "<none>")
+		}
+		for _, name := range names {
+			include := true
+			if params != nil {
+				params.ImageName = name
+			}
+			for _, filter := range filters {
+				include = include && filter(&img, info)
+			}
+
+			if include {
+				newImage := img
+				newImage.Names = []string{name}
+				imagesFiltered = append(imagesFiltered, &newImage)
+			}
 		}
 	}
-	return pullNames, nil
+
+	for _, i := range imagesFiltered {
+		newImages = append(newImages, ir.newFromStorage(i))
+	}
+	return newImages, nil
 }
 
-// pull is a temporary function for stage1 to be able to pull images during the image
-// resolution tests.  it will be replaced in stage2 with a more robust function.
-func (i *Image) pull(name string, writer io.Writer, r *libpod.Runtime) (string, error) {
-	options := libpod.CopyOptions{
-		Writer:              writer,
-		SignaturePolicyPath: r.GetConfig().SignaturePolicyPath,
+// getImageDigest creates an image object and uses the hex value of the digest as the image ID
+// for parsing the store reference
+func getImageDigest(src types.ImageReference, ctx *types.SystemContext) (string, error) {
+	newImg, err := src.NewImage(ctx)
+	if err != nil {
+		return "", err
+	}
+	defer newImg.Close()
+	digest := newImg.ConfigInfo().Digest
+	if err = digest.Validate(); err != nil {
+		return "", errors.Wrapf(err, "error getting config info")
 	}
-	return i.runtime.PullImage(name, options)
+	return "@" + digest.Hex(), nil
 }
 
-// Remove an image
-// This function is only complete enough for the stage 1 tests.
-func (i *Image) Remove(force bool) error {
-	_, err := i.runtime.RemoveImage(i.image, force)
-	return err
+// TagImage adds a tag to the given image
+func (i *Image) TagImage(tag string) error {
+	tags := i.Names()
+	if StringInSlice(tag, tags) {
+		return nil
+	}
+	tags = append(tags, tag)
+	return i.imageruntime.store.SetNames(i.ID(), tags)
+}
+
+// PushImage pushes the given image to a location described by the given path
+func (i *Image) PushImage(destination, manifestMIMEType, authFile, signaturePolicyPath string, writer io.Writer, forceCompress bool, signingOptions SigningOptions, dockerRegistryOptions *DockerRegistryOptions) error {
+	// PushImage pushes the src image to the destination
+	//func PushImage(source, destination string, options CopyOptions) error {
+	if destination == "" {
+		return errors.Wrapf(syscall.EINVAL, "destination image name must be specified")
+	}
+
+	// Get the destination Image Reference
+	dest, err := alltransports.ParseImageName(destination)
+	if err != nil {
+		if hasTransport(destination) {
+			return errors.Wrapf(err, "error getting destination imageReference for %q", destination)
+		}
+		// Try adding the images default transport
+		destination2 := DefaultTransport + destination
+		dest, err = alltransports.ParseImageName(destination2)
+		if err != nil {
+			// One last try with docker:// as the transport
+			destination2 = DefaultTransport + destination
```

Comment:
This is pointless - we're trying the same thing as we just did. This probably used to be "docker://" hardcoded, but once it was changed to DefaultTransport there's no point in it being here

Reply:
That is existing code @mheon as is alot of this stuff.  I think they were trying to handle a case where the defaulttransport is not docker:// ... easy enough to fix.

Reply:
Yeah, I figured it was preexisting. No reason for it to exist after we merged, though :smile:

Reply:
fixed